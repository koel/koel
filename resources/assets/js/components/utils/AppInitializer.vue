<template>
  <slot />
</template>

<script lang="ts" setup>
import { onMounted } from 'vue'
import { useErrorHandler, useOverlay } from '@/composables'
import { commonStore, preferenceStore as preferences } from '@/stores'
import { socketListener, socketService, uploadService } from '@/services'

const emits = defineEmits<{
  (e: 'success'): void
  (e: 'error', err: unknown): void
}>()

const { showOverlay, hideOverlay } = useOverlay()

/**
 * Request for notification permission if it's not provided and the user is OK with notifications.
 */
const requestNotificationPermission = async () => {
  if (preferences.show_now_playing_notification && window.Notification && window.Notification.permission !== 'granted') {
    preferences.show_now_playing_notification = await window.Notification.requestPermission() === 'denied'
  }
}

onMounted(async () => {
  showOverlay({ message: 'Just a little patienceâ€¦' })

  try {
    await commonStore.init()

    await requestNotificationPermission()

    window.addEventListener('beforeunload', (e: BeforeUnloadEvent) => {
      if (uploadService.shouldWarnUponWindowUnload() || preferences.confirm_before_closing) {
        e.preventDefault()
        e.returnValue = ''
      }
    })

    await socketService.init() && socketListener.listen()

    emits('success')
  } catch (error: unknown) {
    useErrorHandler().handleHttpError(error)
    emits('error', error)
  } finally {
    hideOverlay()
  }
})
</script>

<style lang="postcss" scoped>

</style>
