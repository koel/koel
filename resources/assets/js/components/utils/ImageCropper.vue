<template>
  <div class="w-full h-full fixed top-0 left-0 flex items-center justify-center z-[99] bg-black/70">
    <div class="relative max-w-full max-h-full rounded-md flex">
      <Cropper
        ref="cropper"
        :max-width="config.maxWidth"
        :min-width="config.minWidth"
        :src="source"
        :stencil-props="{ aspectRatio: 1 }"
      />
      <div class="fixed top-6 right-6 flex flex-1 gap-2">
        <Btn success @click.prevent="crop">Crop</Btn>
        <Btn transparent @click.prevent="emits('cancel')">Cancel</Btn>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref, toRefs } from 'vue'
import { Cropper } from 'vue-advanced-cropper'
import 'vue-advanced-cropper/dist/style.css'

import Btn from '@/components/ui/form/Btn.vue'

const props = withDefaults(defineProps<{
  source?: string | null
  config?: {
    minWidth: number
    maxWidth?: number
  }
}>(), {
  source: null,
  config: () => ({
    minWidth: 192,
  }),
})

const emits = defineEmits<{
  (e: 'crop', result: string): void
  (e: 'cancel'): void
}>()
const { source, config } = toRefs(props)
const cropper = ref<typeof Cropper>()

const crop = () => emits('crop', cropper.value?.getResult().canvas.toDataURL())
</script>
