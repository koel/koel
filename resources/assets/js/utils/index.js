export * from "./alerts";
export * from "./filters";
export * from "./formatters";
export * from "./supports";
export * from "./common";
export * from "./$";
