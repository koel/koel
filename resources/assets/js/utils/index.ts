export * from './formatters'
export * from './supports'
export * from './common'
export * from './$'
export * from './helpers'
export * from './fileReader'
export * from './directoryReader'
export * from './logger'
export * from './floatingUi'
export * from './eventBus'
