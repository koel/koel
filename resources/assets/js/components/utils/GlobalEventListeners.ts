/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */

import { defineComponent, onMounted } from 'vue'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow } from '@/utils'
import { useDialogBox, useMessageToaster, useRouter } from '@/composables'

export const GlobalEventListeners = defineComponent({
  setup (props, { slots }) {
    let toastSuccess: ReturnType<typeof useMessageToaster>['toastSuccess']
    let showConfirmDialog: ReturnType<typeof useDialogBox>['showConfirmDialog']
    let go: ReturnType<typeof useRouter>['go']

    onMounted(() => {
      toastSuccess = useMessageToaster().toastSuccess
      showConfirmDialog = useDialogBox().showConfirmDialog
      go = useRouter().go
    })

    eventBus.on('PLAYLIST_DELETE', async playlist => {
      if (await showConfirmDialog(`Delete the playlist "${playlist.name}"?`)) {
        await playlistStore.delete(playlist)
        toastSuccess(`Playlist "${playlist.name}" deleted.`)
        go('home')
      }
    }).on('PLAYLIST_FOLDER_DELETE', async folder => {
      if (await showConfirmDialog(`Delete the playlist folder "${folder.name}"?`)) {
        await playlistFolderStore.delete(folder)
        toastSuccess(`Playlist folder "${folder.name}" deleted.`)
        go('home')
      }
    }).on('LOG_OUT', async () => {
      await userStore.logout()
      authService.destroy()
      forceReloadWindow()
    })

    return () => slots.default?.()
  }
})
