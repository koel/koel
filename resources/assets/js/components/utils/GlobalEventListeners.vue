<template>
  <span></span>
</template>

<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, preferenceStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow, requireInjection } from '@/utils'
import { DialogBoxKey, MessageToasterKey, RouterKey } from '@/symbols'

const toaster = requireInjection(MessageToasterKey)
const dialog = requireInjection(DialogBoxKey)
const router = requireInjection(RouterKey)

eventBus.on({
  PLAYLIST_DELETE: async (playlist: Playlist) => {
    if (await dialog.value.confirm(`Delete the playlist "${playlist.name}"?`)) {
      await playlistStore.delete(playlist)
      toaster.value.success(`Playlist "${playlist.name}" deleted.`)
      router.go('home')
    }
  },

  PLAYLIST_FOLDER_DELETE: async (folder: PlaylistFolder) => {
    if (await dialog.value.confirm(`Delete the playlist folder "${folder.name}"?`)) {
      await playlistFolderStore.delete(folder)
      toaster.value.success(`Playlist folder "${folder.name}" deleted.`)
      router.go('home')
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  LOG_OUT: async () => {
    await userStore.logout()
    authService.destroy()
    forceReloadWindow()
  }
})

router.onRouteChanged(() => {
  // Hide the extra panel away if a main view is triggered on mobile.
  if (isMobile.phone) {
    preferenceStore.showExtraPanel = false
  }
})
</script>
