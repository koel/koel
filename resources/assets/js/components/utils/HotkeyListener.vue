<template>
  <GlobalEvents
    @keydown.space="togglePlayback"
    @keydown.j="playNext"
    @keydown.k="playPrev"
    @keydown.f="search"
    @keydown.l="toggleLike"
  />
</template>

<script lang="ts" setup>
import { GlobalEvents } from 'vue-global-events'
import { eventBus } from '@/utils'
import { playbackService, socketService } from '@/services'
import { favoriteStore, queueStore } from '@/stores'

const togglePlayback = (e: KeyboardEvent) => {
  if (
    !(e.target instanceof Document) &&
    (e.target as Element).matches('input, textarea, button, select') &&
    !(e.target as Element).matches('input[type=range]')
  ) {
    return true
  }

  e.preventDefault()
  playbackService.toggle()

  return false
}

/**
 * Play the previous song when user presses K.
 */
const playPrev = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  e.preventDefault()
  playbackService.playPrev()

  return false
}

/**
 * Play the next song when user presses J.
 */
const playNext = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  e.preventDefault()
  playbackService.playNext()

  return false
}

/**
 * Put focus into the search field when user presses F.
 */
const search = (e: KeyboardEvent) => {
  if (
    !(e.target instanceof Document) &&
    (e.target as Element).matches('input, select, textarea') && !(e.target as Element).matches('input[type=range]')
  ) {
    return true
  }

  if (e.metaKey || e.ctrlKey) {
    return true
  }

  e.preventDefault()
  eventBus.emit('FOCUS_SEARCH_FIELD')

  return false
}

/**
 * Like/unlike the current song when use presses L.
 */
const toggleLike = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  if (!queueStore.current) {
    return false
  }

  favoriteStore.toggleOne(queueStore.current)
  socketService.broadcast('SOCKET_SONG', queueStore.current)

  return false
}
</script>
