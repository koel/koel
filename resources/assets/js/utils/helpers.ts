import { isObject } from 'lodash'
import { inject, InjectionKey, isRef, provide, readonly, shallowReadonly } from 'vue'
import { ReadonlyInjectionKey } from '@/symbols'
import { logger } from '@/utils'

export const use = <T> (value: T, cb: (arg: T) => void) => {
  if (typeof value === 'undefined' || value === null) {
    return
  }

  cb(value)
}

export const arrayify = <T> (maybeArray: T | Array<T>) => ([] as Array<T>).concat(maybeArray)

// @ts-ignore
export const noop = () => {
}

export const limitBy = <T> (arr: T[], count: number, offset: number = 0): T[] => arr.slice(offset, offset + count)

export const provideReadonly = <T> (key: ReadonlyInjectionKey<T>, value: T, deep = true, mutator?: Closure) => {
  mutator = mutator || (v => isRef(value) ? (value.value = v) : (value = v))

  if (!isObject(value)) {
    logger.warn(`value cannot be made readonly: ${value}`)
    provide(key, [value, mutator])
  } else {
    provide(key, [deep ? readonly(value) : shallowReadonly(value), mutator])
  }
}

export const requireInjection = <T> (key: InjectionKey<T>, defaultValue?: T) => {
  const value = inject(key, defaultValue)

  if (typeof value === 'undefined') {
    throw new Error(`Missing injection: ${key.toString()}`)
  }

  return value
}

export const dbToGain = (db: number) => Math.pow(10, db / 20) || 0
