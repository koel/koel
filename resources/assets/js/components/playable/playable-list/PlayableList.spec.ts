import { ref } from 'vue'
import { describe, expect, it } from 'vitest'
import { createHarness } from '@/__tests__/TestHarness'
import { arrayify } from '@/utils/helpers'
import {
  FilteredPlayablesKey,
  PlayableListConfigKey,
  PlayableListContextKey,
  PlayableListSortFieldKey,
  PlayableListSortOrderKey,
  SelectedPlayablesKey,
} from '@/symbols'
import Component from './PlayableList.vue'

describe('playableList.vue', () => {
  const h = createHarness()

  const renderComponent = async (
    songs: MaybeArray<Playable>,
    config: Partial<PlayableListConfig> = {
      sortable: true,
      reorderable: true,
    },
    context: PlayableListContext = {
      type: 'Album',
    },
    selectedPlayables: Playable[] = [],
    sortField: PlayableListSortField = 'title',
    sortOrder: SortOrder = 'asc',
  ) => {
    songs = arrayify(songs)

    const sortFieldRef = ref(sortField)
    const sortOrderRef = ref(sortOrder)

    const rendered = h.visit('/songs').render(Component, {
      global: {
        stubs: {
          VirtualScroller: h.stub(),
          PlayableListSorter: h.stub(),
          PlayableListHeader: h.stub(),
        },
        provide: {
          [<symbol>FilteredPlayablesKey]: [ref(songs)],
          [<symbol>SelectedPlayablesKey]: [ref(selectedPlayables), (value: Playable[]) => (selectedPlayables = value)],
          [<symbol>PlayableListConfigKey]: [config],
          [<symbol>PlayableListContextKey]: [context],
          [<symbol>PlayableListSortFieldKey]: [sortFieldRef, (value: PlayableListSortField) => (sortFieldRef.value = value)],
          [<symbol>PlayableListSortOrderKey]: [sortOrderRef, (value: SortOrder) => (sortOrderRef.value = value)],
        },
      },
    })

    return {
      ...rendered,
      songs,
    }
  }

  it('renders', async () => {
    const { html } = await renderComponent(h.factory('song', 5))
    expect(html()).toMatchSnapshot()
  })
})
