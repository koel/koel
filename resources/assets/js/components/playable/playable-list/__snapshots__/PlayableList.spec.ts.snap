// Vitest Snapshot v1

exports[`playableList.vue > renders 1`] = `<div class="playable-list-wrap relative flex flex-col flex-1 overflow-auto py-0 px-3 md:p-0" data-testid="song-list" tabindex="0"><br data-testid="stub" content-type="songs"><br data-testid="stub" item-height="64" items="[object Object],[object Object],[object Object],[object Object],[object Object]"></div>`;
